src/Middleware/RequestDecompression/src/RequestDecompressionMiddleware.cs

```diff
@@ -62,7 +62,9 @@ private async Task InvokeCore(HttpContext context, Stream decompressionStream)
                 context.GetEndpoint()?.Metadata?.GetMetadata<IRequestSizeLimitMetadata>()?.MaxRequestBodySize
                     ?? context.Features.Get<IHttpMaxRequestBodySizeFeature>()?.MaxRequestBodySize;
 
-            context.Request.Body = new SizeLimitedStream(decompressionStream, sizeLimit);
+            context.Request.Body = new SizeLimitedStream(decompressionStream, sizeLimit, (long totalBytesRead, long sizeLimit) => throw new BadHttpRequestException(
+                    $"The request's Content-Length {totalBytesRead} is larger than the request body size limit {sizeLimit}.",
```

Comment:
This error message is imho confusing since the Content-Length actually doesn't exceeed the limit, only the decompressed body does (if the Content-Length already exceeded the limit, it wouldn't even get to this middleware).

Also using `totalBytesRead` here doesn't seem helpful here since it's just a random value that is slightly above the `sizeLimit`. The actual decompressed length is unknown here.

```suggestion
                    $"The decompressed request body is larger than the request body size limit {sizeLimit}.",
```